sdk/core/core/src/az_span.c

```diff
@@ -176,55 +175,43 @@ AZ_NODISCARD az_result az_span_find(az_span source, az_span target, az_span* out
       {
         // All bytes in `target` have been checked and matched the corresponding bytes in `source` (from the start point `i`),
         // so this is indeed an instance of `target` in that position of `source` (step 4.).
-        // Here we create a span in `source` to map the occurrence of `target` in `source` (pardon the redundancy of words, done for explicit clarity).
-        *out_span = az_span_init(source_ptr + i, target_length, target_length);
 
-        return AZ_OK;
+        return i;
       }
     }
   }
 
   // If the function hasn't returned before, all positions 
   // of `source` have been evaluated but `target` could not be found.
-  return AZ_ERROR_ITEM_NOT_FOUND;
+  const int32_t TARGET_NOT_FOUND = -1;
+
+  return TARGET_NOT_FOUND;
 }
 
 AZ_NODISCARD az_span az_span_token(az_span source, az_span delimiter, az_span* out_remainder)
 {
   AZ_PRECONDITION_VALID_SPAN(delimiter, 1, false);
   AZ_PRECONDITION_NOT_NULL(out_remainder);
 
-  if (az_span_is_content_equal(source, AZ_SPAN_NULL))
+  if (az_span_length(source) == 0)
   {
     return AZ_SPAN_NULL;
   }
   else
   {
-    uint8_t* source_ptr = az_span_ptr(source);
-    int32_t source_length = az_span_length(source);
-    int32_t source_capacity = az_span_capacity(source);
+    int32_t index = az_span_find(source, delimiter);
 
-    az_span instance;
-    if (az_span_find(source, delimiter, &instance) == AZ_OK)
+    if (index != -1)
     {
-      uint8_t* instance_ptr = az_span_ptr(instance);
-      int32_t instance_length = az_span_length(instance);
-
-      *out_remainder = az_span_init(
-        instance_ptr + instance_length, 
-        (int32_t)(source_length - instance_length - (instance_ptr - source_ptr)),
-        (int32_t)(source_capacity - instance_length - (instance_ptr - source_ptr)));
-
-      return az_span_init(
-        source_ptr, 
-        (int32_t)(instance_ptr - source_ptr), 
-        (int32_t)(instance_ptr - source_ptr));
+      *out_remainder = az_span_slice(source, index + az_span_length(delimiter), az_span_length(source));
```

Comment:
First off I don't think so. I thought it should be the capacity of the source too but that would essentially be an overflow. 

This might be worth a separate discussion actually. Why does slice allow to extend to the end of capacity? Shouldn't it be `az_span_length()`?

Reply:
I would say the capacity should be preserved somehow, don't you think?
For the usage we have for az_span_token, only the ptr and length are relevant though.

Reply:
In this case I think it should be kept as `az_span_length()`. Example here:

```c
buf[50];
az_span example = az_span_init(buf, 0, sizeof(buf));
topic_get(&example);
//Results in az/iot/hub/az_iot_hub_blah/?key=value
//Length = 37 , Capacity = 50

//Issue it again with same span but no props
topic_get(&example);
//Results in "az/iot/hub/az_iot_hub_blah/"
//Length = 27 , Capacity = 50

az_span_find(&example, "blah/");
//Returns the index but the returned "remainder" will include 
// the old props if using capacity which we don't want.
```

Reply:
> //Returns the index but the returned "remainder" will include 
// the old props if using capacity which we don't want.

I like the approach of reasoning through the behavior by looking at the example usage. Can you update your example to use `az_span_toke`n, though, because `find` now doesn't have a `remainder` and the caller  can slice however they wish (keep capacity, or keep length, depending on  what they want to do next with the span).

It is not clear that having the remainder capacity contain the previous data is a bad thing. It's the same underlying buffer, and you may want to append stuff to it (to overwrite whatever was there before).

Reply:
Thanks for the correction. Updated the code snippet above to use `token`

Reply:
I think maintaining the capacity here makes most sense to me (so if I were to pick one for this PR, I would chose that). Rationale: TODO :)

Reply:
To make sure we're on the same page here, the change should NOT be made. 
It should stay as 
```c
*out_remainder = az_span_slice(
    source, index + az_span_length(delimiter), az_span_length(source));
```

The slice API already maintains the capacity here:
https://github.com/Azure/azure-sdk-for-c/blob/88baf33becd27795facc78457ef0e2f070acbf96/sdk/core/core/src/az_span.c#L42-L45

Specifically the last param to `init` being `capacity - start_index`